File: cses/02_sorting_searching/1660_subarray-sums-i/solution.py
import sys


def main():
    # Đọc toàn bộ dữ liệu một lần rồi tách số để tăng tốc (tránh TLE)
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    x = int(data[idx]); idx += 1

    ans = 0
    cur = 0  # tổng của cửa sổ [left, right]
    left = 0
    # Two-pointer: mở rộng right, cộng a[right] vào cur
    for right in range(n):
        cur += int(data[idx + right])
        # Thu hẹp cửa sổ từ bên trái khi tổng vượt quá x
        while cur > x:
            cur -= int(data[idx + left])
            left += 1
        # Cửa sổ hiện tại có tổng đúng bằng x => đếm thêm một subarray
        if cur == x:
            ans += 1

    sys.stdout.write(str(ans) + "\n")


main()

File: cses/02_sorting_searching/1660_subarray-sums-i/solution.cpp
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    // Đọc n và tổng mục tiêu target (dùng long long vì tổng có thể tới 2*10^14)
    int n;
    long long target;
    std::cin >> n >> target;

    std::vector<long long> values(n);
    for (long long& value : values) {
        std::cin >> value;
    }

    long long answer = 0;
    long long current_sum = 0;  // tổng của cửa sổ [left, right]
    int left = 0;

    // Two-pointer: mở rộng right, cộng dồn vào current_sum
    for (int right = 0; right < n; ++right) {
        current_sum += values[right];
        // Thu hẹp cửa sổ từ bên trái khi tổng vượt quá target
        while (current_sum > target) {
            current_sum -= values[left];
            ++left;
        }
        // Cửa sổ hiện tại có tổng đúng bằng target => đếm thêm một subarray
        if (current_sum == target) {
            ++answer;
        }
    }

    std::cout << answer << '\n';
    return 0;
}
